utils/src/dyn_error.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+
+/// `DynError` is a sub-trait of `Error` that implements `Eq`. Also, `Box<dyn DynError>`
+/// implements `Clone`.
+pub trait DynError: std::error::Error + Send + Sync {
+    fn as_any(&self) -> &dyn Any;
+
+    fn dyn_eq(&self, other: &dyn Any) -> bool;
+
+    fn dyn_clone(&self) -> Box<dyn DynError>;
+}
+
+impl<T: 'static + std::error::Error + Eq + Clone + Send + Sync> DynError for T {
```

Comment:
Why not `<'a, T: 'a, ...>` instead of `<T: 'static, ...>`?

Reply:
> Why not `<'a, T: 'a, ...>` instead of `<T: 'static, ...>`?

I don't think that would work without introducing a lifetime parametrisation to the trait or the type.

Reply:
Yeah, it was needed because `Any` needs `'static`. This was has been removed though.